docs/design/federated-services.md

```diff
@@ -0,0 +1,514 @@
+# Kubernetes Cluster Federation (a.k.a. "Ubernetes")
+## Cross-cluster Load Balancing and Service Discovery 
+### Requirements and System Design
+### by Quinton Hoole, Dec 3 2015
+
+## Requirements
+
+### Discovery, Load-balancing and Failover
+
+1. **Internal discovery and connection**: Pods/containers (running in
+   a Kubernetes cluster) must be able to easily discover and connect
+   to endpoints for Kubernetes services on which they depend in a
+   consistent way, irrespective of whether those services exist in a
+   different kubernetes cluster within the same cluster federation.
+   Hence-forth referred to as "cluster-internal clients", or simply
+   "internal clients".
+1. **External discovery and connection**: External clients (running
+   outside a Kubernetes cluster) must be able to discover and connect
+   to endpoints for Kubernetes services on which they depend.
+   1. **External clients predominantly speak HTTP(S)**: External
+      clients are most often, but not always, web browsers, or at
+      least speak HTTP(S) (TBD list notable exceptions.  DNS servers?
+      SIP servers?  Database Servers?)
```

Comment:
Message buses are the most common "cross cluster" enterprise case I'm familiar with that need to be exposed via TCP to the edge, although it's almost always Java and almost always TLS, so SNI-aware edge routers can solve some of this.